patches/api/0246-Registry-Modification-API.patch

```diff
@@ -0,0 +1,703 @@
++    public static final RegistryKey<TrimMaterial> TRIM_MATERIAL = create("trim_material", TrimMaterial.class);
++    public static final RegistryKey<TrimPattern> TRIM_PATTERN = create("trim_pattern", TrimPattern.class);
++
++    private final Key key;
++    private final Class<T> type;
++
++    private RegistryKey(final Key key, final Class<T> type) {
++        this.key = key;
++        this.type = type;
++    }
++
++    public @NotNull Key key() {
++        return this.key;
++    }
++
++    public @NotNull Class<T> type() {
++        return this.type;
++    }
++
++    @Override
++    public String toString() {
++        return "RegistryKey[" + this.key + " / " + this.type.getSimpleName() + "]";
++    }
++
++    static <T extends Keyed> RegistryKey<T> create(@Subst("some_key") final String key, final Class<T> type) {
++        final RegistryKey<T> registryKey = new RegistryKey<>(Key.key(Key.MINECRAFT_NAMESPACE, key), type);
++        REGISTRY_KEYS.add(registryKey);
++        return registryKey;
++    }
++}
+diff --git a/src/main/java/io/papermc/paper/registry/RegistryManager.java b/src/main/java/io/papermc/paper/registry/RegistryManager.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..bba80f38b9a1ebaf06161332e369d227ec46f7c6
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/registry/RegistryManager.java
+@@ -0,0 +1,39 @@
++package io.papermc.paper.registry;
++
++import io.papermc.paper.registry.event.listener.RegistryAdditionListener;
++import io.papermc.paper.registry.event.listener.RegistryPreFreezeListener;
++import net.kyori.adventure.key.Keyed;
++import org.jetbrains.annotations.ApiStatus;
++import org.jetbrains.annotations.NotNull;
++
++/**
++ * Register listeners to points in a registry's lifecycle.
++ */
++@ApiStatus.NonExtendable
++@ApiStatus.Experimental
++public interface RegistryManager {
++
++    /**
++     * Register a listener which is called when an object is about
++     * to be added to the specified registry. Useful for modifying
++     * objects before they are added.
++     *
++     * @param registryKey the registry key
++     * @param listener an instance of the listener
++     * @param <T> object type
++     * @param <B> object builder type
++     */
++    <T extends Keyed, B extends RegistryBuilder<T>> void registerAdditionListener(@NotNull RegistryKey<T> registryKey, @NotNull RegistryAdditionListener<T, B> listener);
++
++    /**
++     * Register a listener which is called when a registry is about
++     * to be frozen. Useful for adding new entries to registries
++     * that support custom entries.
++     *
++     * @param registryKey the registry key
++     * @param listener an instance of the listener
++     * @param <T> object type
++     * @param <B> object builder type
++     */
++    <T extends Keyed, B extends RegistryBuilder<T>> void registerPreFreezeListener(@NotNull RegistryKey<T> registryKey, @NotNull RegistryPreFreezeListener<T, B> listener);
++}
```

Comment:
It may just be me, but I find the usage of "register" in the context of registries despite being used for the listener system be a bit confusing. 

Although it would deviate from PluginManger's registerEvent naming scheme.. I think it may be better to reduce confusion.
Maybe something like:
- ``newRegisterListener``, ``newRegistrationListener``, ``onRegistration``
- ``onFinalization``, ``onPreFreeze``, ``onRegistryFinalization``

I am not sure, the point is, I think that we should perhaps not use the term register here as it may be a bit confusing. Although, just seeing what others might think here.